### PR TITLE
Script parser

### LINKED ISSUE #976: Create statement parser for conditionals in pipeline configurations
**Is your feature request related to a problem? Please describe.**
Currently there is no standard way to convert a string representing conditional statement written in a script like syntax.

**Describe the solution you'd like**
Create a script parser class that converts a conditional statement string to a parse tree for evaluation. The parser should be generated by ANTLR based on the syntax outlined in [RFC #522](https://github.com/opensearch-project/data-prepper/issues/522).

**Describe alternatives you've considered (Optional)**

1. JavaCC
2. Custom Parser
3. SpEL
4. GOLDEngine

**Additional context**

### PR DESCRIPTION
**Pull request by sbayer55**
### Description
Added data-prepper-script package.
Added Antlr grammer file for DataPrepperScript (when statement) syntax from [#522](https://github.com/opensearch-project/data-prepper/issues/522)
Added skeleton Interface and Class for Script engine. To be used by Data Prepper Core and Plugins to evaluate conditional statements.
 
### Issues Resolved
#976
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-script/build.gradle [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    id 'idea'
+}
+
+group 'org.opensearch.dataprepper'
+version '1.3.0-SNAPSHOT'
+
+repositories {
+    mavenCentral()
+}
+
```
Comment (graytaylor0) [inline data-prepper-script/build.gradle:18]: I think this can all be removed since the parent project provides it. Same for the `test` section

File: data-prepper-script/src/main/antlr/DataPrepperScript.g4 [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+fragment
+JsonPointerStringCharacter
+    : ~["\\]
+    | JsonPointerEscapeSequence
+    ;
+
+fragment
+JsonPointerCharacter
+    : [A-Za-z0-9_]
```
Comment (cmanning09) [inline data-prepper-script/src/main/antlr/DataPrepperScript.g4:74]: Here is a valid key that is not supported by this syntax: `a1K.k3-01_02`

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/ScriptEngine.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+     *
+     * @param statement string to be parsed and evaluated
+     * @param context event used to resolve external references in the statement
+     * @param toValueType target class for coercion of result object
+     * @param <T> target class type for coercion of result object
+     * @return coerced result of statement evaluation
+     *
+     * @throws ClassCastException when unable to course statement result to type T
+     */
+    <T> T evaluate(final String statement, final Event context, final Class<T> toValueType) throws ClassCastException;
```
Comment (cmanning09) [inline data-prepper-script/src/main/java/org/opensearch/dataprepper/script/ScriptEngine.java:39]: Why do we need to overload evaluate? Can wee just use the last one? If we need a generic option the client can call evaluate with the `toValueType` set to `Object.class`

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/ScriptParser.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    private final DataPrepperScriptLexer lexer;
+    private final DataPrepperScriptParser parser;
+
+    public ScriptParser() {
+        lexer = new DataPrepperScriptLexer(EMPTY_STREAM);
+
+        final CommonTokenStream tokenStream = new CommonTokenStream(lexer);
+        parser = new DataPrepperScriptParser(tokenStream);
+    }
```
Comment (cmanning09) [inline data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/ScriptParser.java:36]: We should be using dependency injection.

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * @since 1.3
+ * LogListener is a utility listener that logs every event enter and exit. Useful for debugging and developing new
+ * listeners.
```
Comment (cmanning09) [inline data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java:25]: This seems beyond the scope of the PR. What is the benefit of this logger over using a debugger? Might I propose, we separate the LogListeners out of this PR?

File: data-prepper-script/src/test/java/org/opensearch/dataprepper/script/parser/ScriptParserTest.java [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        scriptParser = new ScriptParser();
+        listener = new TestListener();
+        walker = new ParseTreeWalker();
+    }
+
+    private void parseStatement(final String statement) {
+        final ParseTree parseTree = scriptParser.parse(statement);
+        walker.walk(listener, parseTree);
+
+        LOG.info("JSON:\n{}", listener.toPrettyString());
```
Comment (cmanning09) [inline data-prepper-script/src/test/java/org/opensearch/dataprepper/script/parser/ScriptParserTest.java:40]: I am concerned with this testings approach. TestListener has no unit tests at the moment and we are relying on it to validate code. A bug could be introduced in ScriptParser and hidden by TestListener. 

Would it be easier to just validate the parseTree object itself over providing test translating layer that we should validate as well?

File: data-prepper-script/src/test/java/org/opensearch/dataprepper/script/parser/ScriptParserTest.java [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+package org.opensearch.dataprepper.script.parser;
+
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.antlr.v4.runtime.tree.ParseTreeWalker;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.function.Executable;
+import org.opensearch.dataprepper.script.parser.util.ListenerMatcher;
+import org.opensearch.dataprepper.script.parser.util.TestListener;
```
Comment (graytaylor0) [inline data-prepper-script/src/test/java/org/opensearch/dataprepper/script/parser/ScriptParserTest.java:14]: These tests at a high level look good, and I can see that having a `TestListener` makes testing the `ScriptParser` a lot easier. However, it is a little concerning to me that we are heavily relying on two classes here to test the `ScriptParser`, yet we have no isolated testing for the `TestListener` or `ListenerMatcher`, which contain a significant amount of code. Even though `TestListener` is sort of getting tested here when we compare the final results, is there anyway to isolate the testing of `LogListener` somewhere? How do we know for sure that these two classes are performing perfectly for these tests?

**Review by cmanning09 — COMMENTED** [review]
Thanks for putting this together. I got half way through but I wanted to start a conversation with you as I have some high level questions.

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+ */
+public class LogListener implements DataPrepperScriptListener {
```
Comment (graytaylor0) [inline data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java:27]: This is similar to my other comment on `TestListener` and `ListenerMatcher`, but there is a lot of code here in `LogListener`. How do we know it is actually working as intended without having thorough, isolated tests for it?

File: data-prepper-script/src/test/java/org/opensearch/dataprepper/script/parser/ScriptParserTest.java (hunk 6 above)
Comment (graytaylor0) [inline data-prepper-script/src/test/java/org/opensearch/dataprepper/script/parser/ScriptParserTest.java:40]: This seems like a lot of logging, especially at the info level. Is this necessary?

File: data-prepper-script/build.gradle [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+dependencies {
+    antlr 'org.antlr:antlr4:4.9.2'
+    implementation project(':data-prepper-api')
+    implementation 'com.fasterxml.jackson.core:jackson-databind:2.13.1'
+    testImplementation "org.hamcrest:hamcrest:2.2"
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.7.2'
```
Comment (dlvenable) [inline data-prepper-script/build.gradle:24]: Both JUnit dependencies are added by the parent project. You should remove these to keep the version numbers in-sync.

File: data-prepper-script/build.gradle [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+version '1.3.0-SNAPSHOT'
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    antlr 'org.antlr:antlr4:4.9.2'
+    implementation project(':data-prepper-api')
+    implementation 'com.fasterxml.jackson.core:jackson-databind:2.13.1'
```
Comment (dlvenable) [inline data-prepper-script/build.gradle:22]: Jackson versions are defined by the parent project. You should remove the version from this `implementation` to use the parent version.

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/ScriptEngine.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+package org.opensearch.dataprepper.script;
+
+import com.amazon.dataprepper.model.event.Event;
+
+/**
+ * @since 1.3
+ * ScriptEngine interface to abstract the parse and evaluate implementations.
+ */
+public interface ScriptEngine {
```
Comment (dlvenable) [inline data-prepper-script/src/main/java/org/opensearch/dataprepper/script/ScriptEngine.java:14]: This seems to be quite a bit of scope creep. And it adds security implications which we may not want to add.

The current goal is to support expression evaluation. Maybe there would eventually be overlap with scripting, but that overlap need not exist.

I'd like to use a name that is specifically tied to expressions and expression evaluation.

File: data-prepper-script/build.gradle [hunk 12]
```diff
@@ -0,0 +1,30 @@
+/*
```
Comment (dlvenable) [inline data-prepper-script/build.gradle:1]: In line with my comment below, it may be good to rename this to `data-prepper-expression` to limit scope to expressions.

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java [hunk 13]
```diff
@@ -0,0 +1,230 @@
+/*
```
Comment (dlvenable) [inline data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java:1]: I agree with the comments below.

Additionally, is there a reason this is in `main` and not `test`?

File: data-prepper-script/build.gradle [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+plugins {
+    id 'antlr'
+    id 'java'
+    id 'idea'
+}
+
+group 'org.opensearch.dataprepper'
```
Comment (dlvenable) [inline data-prepper-script/build.gradle:12]: Both `group` and `version` should be removed, to clarify @cmanning09 's comment.

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java (hunk 8 above)
Comment (dlvenable) [inline data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java:27]: Where is `DataPrepperScriptListener` defined? I couldn't find it in this PR.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-script/src/main/antlr/DataPrepperScript.g4 (hunk 2 above)
Comment (cmanning09) [reply data-prepper-script/src/main/antlr/DataPrepperScript.g4:74]: Okay, Let's confirm and align with our usage of a pointer which do/don't require escaped json pointers to create a consistent user experience across processors.

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java (hunk 8 above)
Comment (chenqi0805) [reply data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java:27]: +1

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+        if (ctx instanceof TerminalNode) {
+            return Collections.singletonList((TerminalNode) ctx);
+        }
+        else if (ctx.children == null) {
+            return Collections.emptyList();
+        }
+        else {
+            return ctx.children.stream()
+                    .flatMap(tree -> {
+                        if (tree instanceof TerminalNode) {
```
Comment (chenqi0805) [inline data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java:59]: Isn't this already included in L50?

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java (hunk 8 above)
Comment (chenqi0805) [reply data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java:27]: Confirmed that `org.opensearch.dataprepper.script.antlr.*` is auto-generated code.

File: data-prepper-script/src/main/antlr/DataPrepperScript.g4 (hunk 2 above)
Comment (sbayer55) [reply data-prepper-script/src/main/antlr/DataPrepperScript.g4:74]: I opened an RFC to update syntax specifics #1005

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+     * @since 1.3
+     * Get self or child nodes of type Terminal Node (Indivisible Rule/Token)
+     * @param ctx Current parsing context
+     * @return List of self or all child nodes that are Terminal Nodes. List may be empty if no terminal nodes found.
+     */
+    private List<TerminalNode> getTerminalNodes(final ParserRuleContext ctx) {
+        if (ctx instanceof TerminalNode) {
+            return Collections.singletonList((TerminalNode) ctx);
+        }
+        else if (ctx.children == null) {
```
[the remark this answers is not in the document]
Comment (sbayer55) [reply data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java:53]: If there is a syntax error in the statement being parsed a node that should have children may not. This is used to avoid throwing a null pointed exception in those cases.

File: data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java (hunk 15 above)
Comment (sbayer55) [reply data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java:59]: Here tree is a child of ctx

**Comment by sbayer55** [comment]
Closed PR to break down code changes into smaller PRs
